### PR TITLE
networked-hand-controls, with gestures; also, many examples updates

### PR DESCRIPTION
**Pull request by kylebakerio**
Based on the A-Frame built-in `hand-controls` component. Heavily refactored for readability, performance, and functionality. Will likely make an upstream pull request to A-Frame based on these updates in the near future.

Of course, added the ability to integrate with NAF as well. Gesture updates are stored as a schema prop, and so are easy to propagate via NAF. tracked controller demo has been updated to use this component, as well. dist file has been compiled with the new component, which is necessary to be able to use the file without duplicating it.

### CODE REVIEW

**Comment by kylebakerio** [comment]
To test locally, I recommend using the webxr chrome extension. Also, though, feel free to ping me to try it out.

**Comment by kylebakerio** [comment]
"eslint": "^6.6.0", in package.json, I presume that's what we're using? [eslint added support](https://stackoverflow.com/questions/61628947/eslint-optional-chaining-error-with-vscode) for conditional chaning in 7.5

File: src/components/networked-hand-controls.js [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+    
+    // local hands are controlled by one's controllers
+    // non-local hands are controlled by updates via NAF
+    local: false,
+    checkLocalHand() {
+      // this.local = this.el.components.networked.data.owner == NAF.clientId;
+      // ^while this would be more ideal, NAF doesn't populate the owner until later
+      // and we need to know this at init(), and we can't supply schema options
+      // when instantiating a template instance--so, we require adding a class for now.
+      this.el.classList.forEach(aClass => {
```
Comment (vincentfretin) [inline src/components/networked-hand-controls.js:143]: `classList` has a `contains` api that may simplify this code to
`this.local = this.el.classList.contains("local-naf-hand")`
https://developer.mozilla.org/en-US/docs/Web/API/Element/classList

File: src/components/networked-hand-controls.js [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+        this.removeEventListeners();
+      }
+    },
+  
+    tick(time, delta) {
+      if (this.getMesh()?.mixer) this.getMesh().mixer.update(delta / 1000);
+    },
+    
+    update(oldData) {   
+      console.log(oldData, this.data)
```
Comment (vincentfretin) [inline src/components/networked-hand-controls.js:113]: console.log to be removed

File: src/components/networked-hand-controls.js [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+          
+      this.loader = new THREE.GLTFLoader();
+      this.loader.setCrossOrigin('anonymous');
+      
+      const handmodelUrl = this.MODEL_BASE + this.MODEL_URLS[this.data.handModelStyle + this.data.hand.charAt(0).toUpperCase() + this.data.hand.slice(1)];
+  
+      this.loader.load(this.data.handModelURL || handmodelUrl, 
+                       gltf => {
+        const newMesh = gltf.scene.children[0];
+        // const mesh = gltf.scene.children[0];
```
Comment (vincentfretin) [inline src/components/networked-hand-controls.js:64]: commented line to be removed

File: src/components/networked-hand-controls.js [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+      
+      this.checkLocalHand();
+      if (this.local) {
+        Object.keys(this.buttonEventMap).forEach(evtName => {
+          this.eventFunctionMap[this.data.hand][evtName] = this.handleButton.bind(this, ...this.buttonEventMap[evtName])
+        })      
+        
+        this.el.addEventListener('controllerconnected', () => {
+          this.el.setAttribute('networked-hand-controls', 'visible', true);
+          this.el.object3D.visible = true;
```
Comment (vincentfretin) [inline src/components/networked-hand-controls.js:45]: This line may be unnecessary if your update() method is already handling this.

**Review by vincentfretin — CHANGES_REQUESTED** [review]
That's really cool! I tested locally with the PC and the Quest by defining the certificate for https in `server/easyrtc-server.js`. I couldn't make the WebXR Chrome extension work.

Last time I resaved all the examples from vscode with the prettier config in the repo. Can you please open the html file in vscode and resave?

For conditional chaining, you can try updating eslint, be sure there is no error when you run `npm run lint`. Otherwise remove your usage of conditional chaining and replace it by a if.

Please add some documentation about the component in the README and address my comments and it will be good to merge. Thanks for this contribution!

**Comment by vincentfretin** [comment]
Capture d’écran de 2022-08-08 11-51-44
Capture d’écran de 2022-08-08 11-51-51

File: src/components/networked-hand-controls.js (hunk 1 above)
Comment (kylebakerio) [reply src/components/networked-hand-controls.js:143]: I tried to use that, but it wasn't available here. I think A-Frame doesn't pass it along? I'll try again and see if I had some random weirdness going on, but that was my logic here.

File: src/components/networked-hand-controls.js (hunk 2 above)
Comment (kylebakerio) [reply src/components/networked-hand-controls.js:113]: thanks, will do

File: src/components/networked-hand-controls.js (hunk 4 above)
Comment (kylebakerio) [reply src/components/networked-hand-controls.js:45]: Great point! Original component had this, I later added the update handler for NAF, didn't catch the overlap.

File: src/components/networked-hand-controls.js (hunk 1 above)
Comment (kylebakerio) [reply src/components/networked-hand-controls.js:143]: Seems to work now, dunno what was happening at the moment I was coding it. Updated.

**Comment by kylebakerio** [comment]
Sorry for the scope creep... As I went to write out how to implement the new component, I was dissatisfied and realized the interface needed to be improved. In the process, I went ahead and cleaned up some other stuff in the examples section as well.

- added requested readme updates, with documentation and guide on adding tracked controllers
- updated readme list of examples to include video, and to separate updated/non-updated examples, and updated audio glitch link
- updated basic example by pulling relevant sections from the previous version of tracked-controlers.html
- simplified color tracking in basic example
- vastly simplified tracked-controllers.html to be a minimal example that only demonstrates how to add tracked controllers--so, moved all basic comments to the basic example, and removed the nametag syncing functionality
- updated networked-hand-controls to respect color specified in schema, which has been broken in hand-controls component
- updated networked-hand-controls to have a _much_ simpler user interface--it now updates the NAF template cache in a different way that avoids awkward workarounds within the HTML
- added separate nametag example, and added random-color-update button;

**Comment by vincentfretin** [comment]
Most of the changes are fine. I added comments directly on commit https://github.com/networked-aframe/networked-aframe/commit/28dc8995321c278a292cf4acb6604095fe65f8a4 but it doesn't show up here in the PR.

**Comment by vincentfretin** [comment]
You should remove the player-info component in basic.html, it's not used.

**Comment by vincentfretin** [comment]
Please add the new nametag example in index.html

**Review by vincentfretin — APPROVED** [review]
I think it's ready to be merged now. Or do you still need to work on it?

**Comment by kylebakerio** [comment]
I agree, ready when you are.

**Comment by kylebakerio** [comment]
remember to go update the naf-examples glitch! cheers, thanks for reviewing all that.

**Pull request merged**